exp_test.go

```diff
@@ -0,0 +1,228 @@
+				},
+				expectedPackedString: "011040000000000000000000000000000000164242424242424242",
+			},
+			{
+				name: "struct with *string type and no value",
+				input: struct {
+					MTI                  string  `index:"0"`
+					PrimaryAccountNumber *string `index:"2"`
+				}{
+					MTI: "0110",
+				},
+				expectedPackedString: "011000000000000000000000000000000000",
+			},
+			{
+				name: "struct with *string type, no value and keepzero tag",
+				input: struct {
+					MTI                  string  `index:"0"`
+					PrimaryAccountNumber *string `index:"2,keepzero"`
+				}{
+					MTI: "0110",
+				},
+				expectedPackedString: "011000000000000000000000000000000000",
+			},
+
+			// Tests for int type
+			{
+				name: "struct with int type and value set",
+				input: struct {
+					MTI                  string `index:"0"`
+					PrimaryAccountNumber int    `index:"2"`
+				}{
+					MTI:                  "0110",
+					PrimaryAccountNumber: panInt,
+				},
+				expectedPackedString: "011040000000000000000000000000000000164242424242424242",
+			},
+			{
+				name: "struct with int type and no value",
+				input: struct {
+					MTI                  string `index:"0"`
+					PrimaryAccountNumber int    `index:"2"`
+				}{
+					MTI: "0110",
+				},
+				expectedPackedString: "011000000000000000000000000000000000",
+			},
+			{
+				name: "struct with int type, no value and keepzero tag",
+				input: struct {
+					MTI                  string `index:"0"`
+					PrimaryAccountNumber int    `index:"2,keepzero"`
+				}{
+					MTI: "0110",
+				},
+				expectedPackedString: "011040000000000000000000000000000000010",
+			},
+
+			// Tests for *int type
+			{
+				name: "struct with *int type and value set",
+				input: struct {
+					MTI                  string `index:"0"`
+					PrimaryAccountNumber *int   `index:"2"`
+				}{
+					MTI:                  "0110",
+					PrimaryAccountNumber: &panInt,
+				},
+				expectedPackedString: "011040000000000000000000000000000000164242424242424242",
+			},
+			{
+				name: "struct with *int type and no value",
+				input: struct {
+					MTI                  string `index:"0"`
+					PrimaryAccountNumber *int   `index:"2"`
+				}{
+					MTI: "0110",
+				},
+				expectedPackedString: "011000000000000000000000000000000000",
+			},
+			{
+				name: "struct with *int type, no value and keepzero tag",
+				input: struct {
+					MTI                  string `index:"0"`
+					PrimaryAccountNumber *int   `index:"2,keepzero"`
+				}{
+					MTI: "0110",
+				},
+				expectedPackedString: "011000000000000000000000000000000000",
+			},
+		}
+
+		for _, tt := range tests {
+			t.Run(tt.name, func(t *testing.T) {
+				message := iso8583.NewMessage(specs.Spec87ASCII)
+				err := message.Marshal(tt.input)
+				require.NoError(t, err)
+
+				packed, err := message.Pack()
+				require.NoError(t, err)
+
+				require.Equal(t, tt.expectedPackedString, string(packed))
+			})
+		}
+	})
+
+	t.Run("unpack", func(t *testing.T) {
```

Comment:
I would review `Unmarshal`code and write tests to cover all `if` and `case` blocks such as:
* test for setting []byte

and also test negative scenarios:
* using map
* use slice of strings
* etc.